### ISSUE TITLE
Authentication Time Out with TP-Link Archer T2UH on ARCH 4.10.11

### ISSUE THREAD

**Issue by Dumonu** [ulli-kroll/mt7610u#26]
After much difficulty, I managed to find your driver and install it, uninstalling the "official" driver. I no longer crash the module when I try to scan! In fact, it now scans properly! Finally!

But, for some reason, attempting to connect with wpa_supplicant to MYNETWORK doesn't work, causing Authentication time outs. MYNETWORK is secured with WPA2 and I would rather not change router settings.

Here are the contents of my /etc/wpa_supplicant/wpa_supplicant.conf: (SSID and psk changed)
```
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=wheel

network={
    ssid="MYNETWORK"
    proto=RSN
    scan_ssid=1
    key_mgmt=WPA-PSK
    #psk="secretcode"
    psk=00000000000000000000000000000000000000000000000000000000
}
```

Here is what I get when I run wpa_supplicant: (MAC address and SSID changed)
```
[root@jacarch ~]# wpa_supplicant -i wlan0 -c /etc/wpa_supplicant/wpa_supplicant.conf
Successfully initialized wpa_supplicant
wlan0: Trying to associate with 00:00:00:00:00:00 (SSID='MYNETWORK' freq=2462 MHz)
wlan0: Authentication with 00:00:00:00:00:00 timed out.
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: Trying to associate with 00:00:00:00:00:00 (SSID='MYNETWORK' freq=2462 MHz)
wlan0: Authentication with 00:00:00:00:00:00 timed out.
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SSID-TEMP-DISABLED id=0 ssid=“MYNETWORK” auth_failures=1 duration=10 reason=CONN_FAILED
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-SSID-REENABLED id=0 ssid=“MYNETWORK”
wlan0: Trying to associate with 00:00:00:00:00:00 (SSID='MYNETWORK' freq=2462 MHz)
wlan0: Authentication with 00:00:00:00:00:00 timed out.
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SSID-TEMP-DISABLED id=0 ssid=“MYNETWORK” auth_failures=2 duration=20 reason=CONN_FAILED
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-REGDOM-CHANGE init=BEACON_HINT type=UNKNOWN
wlan0: CTRL-EVENT-SSID-REENABLED id=0 ssid=“MYNETWORK”
wlan0: Trying to associate with 00:00:00:00:00:00 (SSID='MYNETWORK' freq=2462 MHz)
wlan0: Authentication with 00:00:00:00:00:00 timed out.
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SSID-TEMP-DISABLED id=0 ssid=“MYNETWORK” auth_failures=3 duration=30 reason=CONN_FAILED
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-SSID-REENABLED id=0 ssid=“MYNETWORK”
wlan0: Trying to associate with 00:00:00:00:00:00 (SSID='MYNETWORK' freq=2462 MHz)
wlan0: Authentication with 00:00:00:00:00:00 timed out.
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SSID-TEMP-DISABLED id=0 ssid=“MYNETWORK” auth_failures=3 duration=30 reason=CONN_FAILED
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
wlan0: CTRL-EVENT-DISCONNECTED bssid=00:00:00:00:00:00 reason=3 locally_generated=1
wlan0: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
^Cnl80211: deinit ifname=wlan0 disabled_11b_rates=0
wlan0: CTRL-EVENT-TERMINATING
```

Is this a driver issue or is it a configuration issue?

**Comment by Dumonu** [comment]
After a lot of analysis, I have determined that the problem is probably in the nl80211 Register Frame Command: `NL80211_CMD_REGISTER_FRAME`. When I run `wpa_supplicant` with the `-d` option, I get a string of these errors:
```
nl80211: Register frame type=0xd0 (WLAN_FC_STYPE_ACTION) nl_handle=0x122f8e0 match=040a
nl80211: Register frame command failed (type=208): ret=-95 (Operation not supported)
nl80211: Register frame match - hexdump(len=2): 04 0a
```
This is followed by a:
```
nl80211: Failed to register Action frame processing - ignore for now
```

I have as of yet not been able to find information on this error. To note, I have found information on a version where Register frame command fails with return -114 (Operation already in progress), but those are caused by attempting to simultaneously run wpa_supplicant and some other wifi connection program.

**Comment by Owersun** [comment]
I have the same problem and I can tell you that this is not the issue.
And moreover I have a dongle working with one wifi network and not working with another. This is some either vendor implementation or setting in WiFi AP that brakes it.

**Comment by Owersun** [comment]
For me it turns out that it is enough to bring the interface up with "ifup wlan0" to bring it to the exactly state described, and then just do the "iwconfig wlan0 essid <SSID>" to assign ssid in the iwconfig. After that everything went through, the association, band selection, channel selection and everything related that allowed correct wifi work to start.

**Comment by ulli-kroll** [comment]
@Dumonu 
@Owersun 
The 802.11 stack in this driver is broken by design (tm)
I won't fix this,
if this driver is ready I can use the mac80211 stack inside the linux kernel.

**Comment by Owersun** [comment]
Arbeiten Sie das an? Oder wir sollten einen anderen Trieber fur T2U suchen? Der arbeitet besser als anderen, und der ist einzig ich kann am 4.x kernel benutzen.

**Comment by ulli-kroll** [comment]
> Arbeiten Sie das an? Oder wir sollten einen anderen Trieber fur T2U suchen? Der arbeitet besser als anderen, und der ist einzig ich kann am 4.x kernel benutzen.
English please ...
What do you mean ?
Fix the timeout or convert to mac80211

**Comment by Owersun** [comment]
Convert to mac802011? I find your repo the most in shape in terms of usability of the module and devices chip based on mt7610u. But I see you complain a lot around that this is a huge pile of garbage that is not compatible with standard linux kernel API's (yeah, nobody argues that), and most your answers to people questions is that explanation. It's not quite clear for me - are you trying to fix that up and just complaining that this is a backpain to do, so people should be more patient, or you are just clarifying that this will never be okey, exactly because the driver is a dump?
Don't get me wrong, I'm not in any way complaining, just trying to understand should I keep an eye on your repo as the one that is the progress most for this driver, or maybe give up trying and avoid anything mt7610u based?..

**Comment by ulli-kroll** [comment]
FWIW
All I want is a mac80211 compatible driver ..
I use Felix mt76 driver and upstream mt7601u driver as blueprint to do this.

Last couple of days I've tested the channel switch code from mt7612u, which doesn't work to my surprise. Today I've discovered some other place in the code (with some other schematics) to do the bandwidth change.
Both are maybe incompatible, don't know.

I have some more ideas this fix this thing, one is using the mt7601u driver with a little patch.
This will end with a 2.4 GHz only driver, but is a good starting point.

Compared to this the mt7612u device is similar to the mt76x2 in Felix' mt76 source.
No special handling, but without **working** LED.
And the code form mt7610u doesn't work here, tough.

**Comment by Dumonu** [comment]
@Owersun 
Is there an iw command equivalent to `iwconfig wlan0 essid <SSID>`? Arch uses iw rather than iwconfig for wifi connection. `iw dev wlan0 connect <SSID>` doesn't seem to do the same thing.

**Comment by Owersun** [comment]
I wasn't able to find the equivalent.
It seems for me iw and iwconfig operates through different API's, because not even the command-sets are different, but the effect also.
Seems iwconfig pull something iw doesn't. Also looking into wireless ubuntu startup scripts it seems that this is known problem - script that processes wireless-... options in interfaces mentions that there are a lot of ambiguity in drivers and all the commands are going to be tried before the interface is up and after, once more, if the failed before.
So basically it's a mess itself, and only guarantee that a driver complies to linux wireless startup scripts is that driver is upstream.

**Comment by ulli-kroll** [comment]
> It seems for me iw and iwconfig operates through different API's, because not even the command-sets are different, but the effect also.

Yes
AFAICR 
iwconfig uses ioctl's with wireless-ext/cfg80211
iw uses netlink with mac80211/cfg80211

wireless-ext is the oldest API and the intention if this was (at the beginning) to get wifi working
But the developers missed something so they added cfg80211.

In between there is iwpriv, with totally vendor specific commands.

 > So basically it's a mess itself, and only guarantee that a driver complies to linux wireless startup scripts is that driver is upstream.

So you now understand how bad it is to fix these timeouts
;-)

**Comment by Maxattax97** [comment]
[I'm having the exact same issue.](https://askubuntu.com/q/928272/450288)

I'm going to try the prescribed fix by @Owersun which is:
```
ifup wlan0
iwconfig wlan0 essid CoolBox_5G     # in my particular case.
```

I'll return with the results afterward.

**Comment by Maxattax97** [comment]
Hmm... No success.
First off, I replaced `ifup wlan0` with `ifconfig wlan0 up`.
Then I ran `iwconfig wlan0 esside CoolBox_5G && iwconfig wlan0` and received back:
```
wlan0     Ralink STA  ESSID:""  Nickname:""
          Mode:Auto  Frequency=2.412 GHz  Access Point: Not-Associated   
          Bit Rate:1 Mb/s   
          RTS thr:off   Fragment thr:off
          Link Quality=10/100  Signal level:0 dBm  Noise level:0 dBm
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0
```

`dmesg` provided some similar logs to before:
```
[  994.390379] 80211> CFG80211_OpsDisconnect ==>
[  994.390380] 80211> ReasonCode = 3
[  994.426236] ==>mt7610u_chip_onoff(): OnOff:0, Reset= 0, pAd->WlanFunCtrl:0xff000003, Reg-WlanFunCtrl=0xff000003
[  994.429843] receive cmd msg fail(-2)
[  994.429910] tx_kickout_fail_count = 0
[  994.429912] tx_timeout_fail_count = 0
[  994.429912] rx_receive_fail_count = 0
[  997.486662] ==>mt7610u_chip_onoff(): OnOff:1, Reset= 0, pAd->WlanFunCtrl:0xff000000, Reg-WlanFunCtrl=0xff000000
[  997.486872] MACVersion = 0x76502000
[  997.487246] MAC[Ver=0x76502000]
[  997.487250] usb 1-5: loading firmware mt7610u.bin
[  997.492319] usb 1-5: firmware mt7610u.bin loaded
[  997.497995] fw version:0.1.00 
[  997.497997] build:7640
[  997.497997] build time:
[  997.497998] ilm length = 68780(bytes)
[  997.497999] dlm length = 11476(bytes)
[  997.580031] #
[  997.586409] loading fw
[  997.587481] .
[  997.594002] .
[  997.600500] .
[  997.607007] .
[  997.613378] .
[  997.619609] .

[  997.708167] #
[  997.924565] RTMP_TimerListAdd: add timer obj ffffac9f03346728!
[  997.924566] RTMP_TimerListAdd: add timer obj ffffac9f03346798!
[  997.924567] RTMP_TimerListAdd: add timer obj ffffac9f03346808!
[  997.924568] RTMP_TimerListAdd: add timer obj ffffac9f033466b8!
[  997.924569] RTMP_TimerListAdd: add timer obj ffffac9f03346568!
[  997.924570] RTMP_TimerListAdd: add timer obj ffffac9f033465d8!
[  997.924571] RTMP_TimerListAdd: add timer obj ffffac9f032d8a60!
[  997.924572] RTMP_TimerListAdd: add timer obj ffffac9f032c78e0!
[  997.924573] RTMP_TimerListAdd: add timer obj ffffac9f032c7958!
[  997.924573] RTMP_TimerListAdd: add timer obj ffffac9f032d8bd0!
[  997.924575] RTMP_TimerListAdd: add timer obj ffffac9f032d8980!
[  997.924576] RTMP_TimerListAdd: add timer obj ffffac9f032d8b60!
[  997.924587] CfgSetCountryRegion():CountryRegion in eeprom was programmed
[  997.924593] CfgSetCountryRegion():CountryRegion in eeprom was programmed
[  997.924622] cfg_mode=13
[  997.924624] wmode_band_equal(): Band Not Equal!
[  997.925183] 1. Phy Mode = 61
[  997.925184] 2. Phy Mode = 61
[  997.929272] /home/alan/src/mt7610u/chips/mt76x0.c:1708 assert (pAd->TxPower[choffset].Channel == 36)failed
[  997.937018] ERROR!!! 
[  997.937020] E2PROM: WRONG VERSION 0x2, should be 1
[  997.938520] BT Coexistence word [66] = 0000ffff
[  997.939022] 0x24 = 0xffef, 0x0104 = 0x0033ffef
[  997.939023] EEPROM_NIC_CFG1_OFFSET = 0xfd11
[  997.939395] mt76x0_read_tx_alc_info_from_eeprom: EEPROM_MT76x0_TEMPERATURE_OFFSET (0xD1) = 0xf8
[  997.939396] mt76x0_read_tx_alc_info_from_eeprom: TemperatureOffset = 0xfffffff8
[  997.939396] Temperature Tx ALC not enabled
[  997.942801] 3. Phy Mode = 61
[  997.950380] TxPath = 1, RxPath = 1, RFIC=17
[  997.950389] RTMPSetPhyMode: channel is out of range, use first channel=1 
[  997.951770] MCS Set = ff 00 00 00 01
[  997.961729] <==== rt28xx_init, Status=0
[  997.961731] 80211> re-init bands...
[  997.961883] 80211> re-init bands...
[  997.961884] 80211> RFICType = 3
[  997.961885] 80211> Number of channel = 44
[  997.961886] 80211> Number of rate = 12
[  997.961886] 80211> CurTxPower = 0 dBm
[  997.961888] 80211> TxStream = 1
[  997.961889] crda> CFG80211_RegRuleApply ==>
[  997.961890] crda> reset chan/power for 2.4GHz
[  997.961891] Chan 001:        power 0 dBm, DFS 0, DFS Type 0
[  997.961892] Chan 002:        power 0 dBm, DFS 0, DFS Type 0
[  997.961893] Chan 003:        power 0 dBm, DFS 0, DFS Type 0
[  997.961894] Chan 004:        power 0 dBm, DFS 0, DFS Type 0
[  997.961895] Chan 005:        power 0 dBm, DFS 0, DFS Type 0
[  997.961896] Chan 006:        power 0 dBm, DFS 0, DFS Type 0
[  997.961897] Chan 007:        power 0 dBm, DFS 0, DFS Type 0
[  997.961898] Chan 008:        power 0 dBm, DFS 0, DFS Type 0
[  997.961899] Chan 009:        power 0 dBm, DFS 0, DFS Type 0
[  997.961900] Chan 010:        power 0 dBm, DFS 0, DFS Type 0
[  997.961901] Chan 011:        power 0 dBm, DFS 0, DFS Type 0
[  997.961902] Chan 012:        power 0 dBm, DFS 0, DFS Type 0
[  997.961903] Chan 013:        power 0 dBm, DFS 0, DFS Type 0
[  997.961904] Chan 014:        power 0 dBm, DFS 0, DFS Type 0
[  997.961905] crda> reset chan/power for 5GHz
[  997.961906] Chan 036:        power 0 dBm, DFS 0, DFS Type 0
[  997.961907] Chan 038:        power 0 dBm, DFS 0, DFS Type 0
[  997.961908] Chan 040:        power 0 dBm, DFS 0, DFS Type 0
[  997.961909] Chan 044:        power 0 dBm, DFS 0, DFS Type 0
[  997.961910] Chan 046:        power 0 dBm, DFS 0, DFS Type 0
[  997.961911] Chan 048:        power 0 dBm, DFS 0, DFS Type 0
[  997.961912] Chan 052:        power 0 dBm, DFS 1, DFS Type 0
[  997.961913] Chan 054:        power 0 dBm, DFS 1, DFS Type 0
[  997.961914] Chan 056:        power 0 dBm, DFS 1, DFS Type 0
[  997.961915] Chan 060:        power 0 dBm, DFS 1, DFS Type 0
[  997.961916] Chan 062:        power 0 dBm, DFS 1, DFS Type 0
[  997.961917] Chan 064:        power 0 dBm, DFS 1, DFS Type 0
[  997.961918] Chan 100:        power 0 dBm, DFS 1, DFS Type 0
[  997.961919] Chan 104:        power 0 dBm, DFS 1, DFS Type 0
[  997.961920] Chan 108:        power 0 dBm, DFS 1, DFS Type 0
[  997.961921] Chan 112:        power 0 dBm, DFS 1, DFS Type 0
[  997.961922] Chan 116:        power 0 dBm, DFS 1, DFS Type 0
[  997.961923] Chan 118:        power 0 dBm, DFS 1, DFS Type 0
[  997.961924] Chan 120:        power 0 dBm, DFS 1, DFS Type 0
[  997.961925] Chan 124:        power 0 dBm, DFS 1, DFS Type 0
[  997.961926] Chan 126:        power 0 dBm, DFS 1, DFS Type 0
[  997.961927] Chan 128:        power 0 dBm, DFS 1, DFS Type 0
[  997.961928] Chan 132:        power 0 dBm, DFS 1, DFS Type 0
[  997.961929] Chan 134:        power 0 dBm, DFS 1, DFS Type 0
[  997.961930] Chan 136:        power 0 dBm, DFS 1, DFS Type 0
[  997.961931] Chan 140:        power 0 dBm, DFS 1, DFS Type 0
[  997.961932] Chan 149:        power 0 dBm, DFS 0, DFS Type 0
[  997.961933] Chan 151:        power 0 dBm, DFS 0, DFS Type 0
[  997.961934] Chan 153:        power 0 dBm, DFS 0, DFS Type 0
[  997.961935] Chan 157:        power 0 dBm, DFS 0, DFS Type 0
[  997.961936] Chan 159:        power 0 dBm, DFS 0, DFS Type 0
[  997.961937] Chan 161:        power 0 dBm, DFS 0, DFS Type 0
[  997.961939] Chan 165:        power 0 dBm, DFS 0, DFS Type 0
[  997.961939] Chan 167 (frq 5835):     not allowed!
[  997.961940] Chan 169 (frq 5845):     not allowed!
[  997.961941] Chan 171 (frq 5855):     not allowed!
[  997.961942] Chan 173 (frq 5865):     not allowed!
[  997.961943] Chan 184 (frq 4920):     not allowed!
[  997.961944] Chan 188 (frq 4940):     not allowed!
[  997.961944] Chan 192 (frq 4960):     not allowed!
[  997.961945] Chan 196 (frq 4980):     not allowed!
[  997.961946] Chan 208 (frq 6040):     not allowed!
[  997.961947] Chan 212 (frq 6060):     not allowed!
[  997.961947] Chan 216 (frq 6080):     not allowed!
[  997.961948] crda> Number of channels = 47
[  997.963024] 0x1300 = 00064300
[  997.963025] RTMPDrvSTAOpen(1):Check if PDMA is idle!
[  997.963144] RTMPDrvSTAOpen(2):Check if PDMA is idle!
```

**Comment by Owersun** [comment]
ifconfig wlan0 up just enables the card.
The whole thing was around the fact that "ifup wlan0" initializes the card using all the scripts around it, wpa_supplicant and all further inits, and the you just push it forward with missing step doing "iwconfig wlan0 essid <SSID>"
Also I've found that be really unrealiable anyway. For example couple of ups/downs put the card into the state it doesn't really work anymore until reboot.

**Comment by Maxattax97** [comment]
If I can get it to a point where it connects and stays that way so long as I don't touch it, I'm happy. Do you think that (at least in my case) it's a power issue?

I'm no wifi expert, but it looks like from the `dmesg` and other output that it's not getting enough juice... NetworkManager's `wifi.powersave` setting is set to 3 ([see AskUbuntu post](https://askubuntu.com/questions/928272/setting-up-mt7610u-wireless-driver)), but I didn't have much success changing it to 2 which is "no powersaving". I could try it again though; experiment a little more.

**Comment by Owersun** [comment]
For me personally one of the failures I didn't catch for a long time was that my router is from Hungary with HU as crda setting and the dongle I was using is from Russia with RU crda setting. And suprisingly 5Ghz channel router was prefering was completely different from what dongle was ready to work with. I really had to force the router to one of the channels that is available in every country (36 if I remember correctly), so the dongle can start seeing it.
I really have had so many issues with this dongle and 5Ghz only wireless I gave up. Probably that was the only device in my home I've said "screw that, I can live with the cable here".

**Comment by Dumonu** [comment]
I've tried the current master (same issue), the mac branch (same issue), and the vanilla branch (didn't compile). So, I'm going to just have to wait till I get access to a wired connection to update my system (this isn't going to be possible until August). I'm reluctant to get _yet another_ new wifi card because this one works just fine with my Windows install. Next wifi card, I'm going to check whether it ACTUALLY supports Linux and doesn't just say that on the box.

**Comment by Dumonu** [comment]
I want to point out that after coming back to this issue (you know, fresh eyes and all that), I managed to get my wifi card working. I did not use this driver in the end, instead I used [xtknight/mt7610u-linksys-ae6000-wifi-fixes](https://github.com/xtknight/mt7610u-linksys-ae6000-wifi-fixes). This had the same crash on scan end issue that other drivers based on the official driver also had, so I decided to dig around in the source files after noticing a function call as part of the reason for the module crash. Several hours later, I discovered that in `CFG80211OS_ScanEnd()` there was a function call to `cfg80211_scan_done()` in which a boolean is passed to the second argument. The second argument should actually be of type `cfg80211_scan_info*`, which is a pointer to a struct! So it was trying to dereference a boolean, which was causing the crash. In any case, my fix can be found at [Dumonu/mt7610u-linksys-ae6000-wifi-fixes](https://github.com/Dumonu/mt7610u-linksys-ae6000-wifi-fixes).

**Comment by ulli-kroll** [comment]
FWIW
commit 06ac1c134825194b5314ecba4352e36be093dcf2
Author: Hans Ulli Kroll <ulli.kroll@googlemail.com>
Date:   Sat Oct 22 12:50:30 2016 +0200
```
Author: Hans Ulli Kroll <ulli.kroll@googlemail.com>
Date:   Sat Oct 22 12:50:30 2016 +0200

fix crash with kernels >= v4.8

in 1d76250bd34af86c6498fc51e50cab3bfbbeceaa
api has changed for cfg80211_scan_done()
    
Signed-off-by: Hans Ulli Kroll <ulli.kroll@googlemail.com>

diff --git a/os/linux/rt_linux.c b/os/linux/rt_linux.c
index 23b7cc5..81d1639 100644
--- a/os/linux/rt_linux.c
+++ b/os/linux/rt_linux.c
@@ -2761,12 +2761,25 @@ void CFG80211OS_ScanEnd(
        IN void *pCB,
        IN BOOLEAN FlgIsAborted)
 {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0))
+       bool abort = true;
+
+       struct cfg80211_scan_info info = {
+               .aborted = abort,
+       };
+#endif
+
 #ifdef CONFIG_STA_SUPPORT
        CFG80211_CB *pCfg80211_CB = (CFG80211_CB *)pCB;
 
 
        CFG80211DBG(RT_DEBUG_ERROR, ("80211> cfg80211_scan_done\n"));
-       cfg80211_scan_done(pCfg80211_CB->pCfg80211_ScanReq, FlgIsAborted);
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0))
+       cfg80211_scan_done(pCfg80211_CB->pCfg80211_ScanReq, &info);
+#else
+       cfg80211_scan_done(pCfg80211_CB->pCfg80211_ScanReq, FlfIsAborted);
+#endif
+
 #endif /* CONFIG_STA_SUPPORT */
 }

```

**Comment by FlorianBador** [comment]
Just in case anyone has the same problem, I personally had these "reason=3 locally_generated=1" errors because I was trying to execute wpa_supplicant while another one was already running from boot. So make ps auxm | grep wpa to see if you don't already have one running, then use /etc/network/interfaces to configure the network instead.

**Comment by Dumonu** [comment]
I've more or less decided that I'm not going to use this driver and I'm just going to keep the other driver up to date as long as that works for me. Running through the same connection steps using the other driver works. However, I'd be interested to see if anyone having this problem using this driver is actually running two wpa_supplicants.
